File: Search_a_2D_Matrix/solution.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    let r = locateRow(target, matrix, 0, matrix.length - 1);
    let row = r >= 0 ? matrix[r] : null;
    if (row) return searchRow(target, row, 0, row.length - 1);
    return false;
};

var locateRow = function (target, matrix, start, end) {
    if (start > end) return -1;
    
    let mid = Math.floor((start + end) / 2);
    let row = matrix[mid];
    if (row.length <= 0) return -1;
        
    if (target < row[0]) {
        end = mid - 1;
    } else if (target >= row[0]) {
       if (target <= row[row.length-1]) {
           return mid;
       }
        start = mid + 1;
    }
    return locateRow(target, matrix, start, end);
}

var searchRow = function (target, row, start, end) {
    if (start > end) return false;
    
    let mid = Math.floor((start + end) / 2);
    let v = row[mid];
    if (target < v) {
        end = mid - 1;
    } else if (target > v) {
        start = mid + 1;
    } else {
        return true;
    }
    return searchRow(target, row, start, end);
}

File: Search_a_2D_Matrix/solution.py
import math

def searchMatrix(matrix, target):
  COL_LEN = len(matrix[0])
  s = 0
  e = COL_LEN * len(matrix) - 1
  
  while (s <= e):
    m = math.floor((s + e) / 2)
    v = getFromMatrix(m, COL_LEN, matrix)
    if v > target:
      e = m - 1
    elif v < target:
      s = m + 1
    else:
      return True
  return False
  
def getFromMatrix(n, COL_LEN, matrix):
  j = n % COL_LEN
  i = math.floor(n / COL_LEN)
  return matrix[i][j]
